src/core/qgsproviderregistry.cpp

```diff
@@ -525,6 +525,12 @@ void QgsProviderRegistry::registerGuis( QWidget *parent )
   }
 }
 
+void QgsProviderRegistry::registerProvider( QgsProviderMetadata *providerMetadata )
+{
+  if ( providerMetadata )
```

Comment:
Better check for existing providers with the same key here, and if so fail the registration. You could make this return a bool to indicate failures.